### ISSUE TITLE
UI internals (+externals): Think about purging inheritance-based event listeners?

### ISSUE THREAD

**Issue by modeco80** [BespokeSynth/BespokeSynth#800]
## Summary

It seems like the current UI API set which:

 - enforces inheritance
 - enforces implementing/creating an interface for every control signal set
 - requires the "slot" (the derived implementation of aforementioned interface) checking what fired the signal to then detour to internal logic per each callee (callee being the passed in (UI) object)

can be replaced with something like Qt's signal/slot pattern. We could either homebrew this in `bespoke::core` (making the project begin to use namespaces for itself), use Boost.Signals2, or something along those lines.

## Stuff & things

Why place Signal in `bespoke::core` instead of `bespoke::ui` if UI's gonna be the main consumer for now?
- Signals *are* used for UI but could in theory be a pattern used anywhere, which might clean up even more of the codebase (This is how Qt handles it too since there are quite a few non-UI Qt objects, fwiw)
- Doesn't really make sense to put it in UI since it's not implementing anything UI specific
- Signals also reduce vtable weight, and since this issue tackles removing the UI event listener interfaces, may actually completely remove some generated vtables (well, and thin out ones that previously needed to include the listeners. But you get my point)
- Reorganization takes time and thought :(

Very very very simple draft nowhere near any sort of final api for homebrew route:
```cpp
//
// Connect being a template which could take an arbitrary Callable (including the result of std::bind()).
// In this case the Callable passed to it is a lambda which can capture anything
// Return type is something like `bespoke::core::SignalConnection` which can us used to disconnect
//
auto connection = mSomethingImportantButton->Clicked.Connect([&]() {
  // we know we clicked this button specifically, so funnel to (this)->DoSomethingImportant()?
  // for prototype/small code it may even be doable to just implement logic inside of lambdas that are connected
});

// When disconnecting from a signal:
mSomethingImportantButton->Clicked.Disconnect(connection);


// Inside of UI component:

// namespace bespoke::ui {
// class Button : .... {
      core::Signal<void()> Clicked; // Fired when button is clicked
// };
// } // namespace bespoke::ui

// As a complement to above, we could also have methods to hide the signal as an implementation detail,
// but either way works fine.

// When firing somewhere in the component:
Clicked(... /* any varadic arguments compatible with the signal type, or empty for Signal<R()> */);

```
## Motivation/Reasoning

It may require a bit more machinery internally and refactoring of internal/external UI code, but signals are far better and can still be just as statically typed.

Ridding the codebase of interfaces just for UI control signals, as well as also beginning to move to use namespaces are both decent steps towards at least beginning to tackle modernization (if not that, then code cleanup's always a nice thing, right? :stuck_out_tongue_closed_eyes:).

**Comment by awwbees** [comment]
this seems useful, thanks. I am a bit thick, so it will probably take me a while before I am able to spend the time to look at this and really understand it. I am a dinosaur when it comes to keeping up with modern C++

**Comment by modeco80** [comment]
Yeah, if you want I can explain everything in a bit more detail on discord
